system/adb/shell_exec_builtin.c

```diff
@@ -0,0 +1,200 @@
+/****************************************************************************
+ * system/adb/shell_pipe_uv.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <spawn.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#include "adb.h"
+#include "builtin/builtin.h"
+#include "shell_pipe.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int shell_exec_builtin(FAR const char *appname, FAR char * const *argv,
```

Comment:
why not call system or nsh api instead?

Reply:
This is a first iteration regarding NSH integration in "adb shell" command. The technical solution is to leverage pipe for both stdout and stdin streams. The issue is I need a custom function to configure fd 0 and fd 1 for NSH process. That's what shell_exec_builtin() does (based on ./builtin/exec_builtin.c but with more posix_spawn_file_actions_addopen()).

Next step could be to try integrating nshlib directly in adbd thread. It may simplify this logic and save lot of RAM.

Reply:
Yes, you can reference how telnetd integrate with nshlib here:
https://github.com/apache/incubator-nuttx-apps/blob/master/nshlib/nsh_telnetd.c
https://github.com/apache/incubator-nuttx-apps/blob/master/netutils/telnetd/telnetd_daemon.c

Reply:
One more simple example to integrate nsh:
https://github.com/apache/incubator-nuttx-apps/tree/master/examples/pty_test

Reply:
@xiaoxiang781216 thank you for your help. The "adb shell" implementation makes more sense now with dup2().

Reply:
LGTM? @xiaoxiang781216 @spiriou should I?

Reply:
The shell implementation could be based on ptmx like in https://github.com/apache/incubator-nuttx-apps/tree/master/examples/pty_test instead of pipes but this can be improved in another patch

Reply:
@spiriou is this patch ready for merge?

Reply:
@xiaoxiang781216 it's not perfect but this can be merged as it's working fine. I'll improve it later.